### ISSUE TITLE
ClientModel: Release `Azure.Variant` in the Azure.Core package

### ISSUE THREAD

**Issue by heaths** [Azure/azure-sdk-for-net#32978]
# Feature Description
`Azure.Variant` is an implementation of a [tagged union](https://en.wikipedia.org/wiki/Tagged_union).  It provides a type we can use for both primitives and value types to avoid boxing .NET intrinsic types.  For example, in libraries like KeyVault mentioned below, it could be used to hold both a string and an int without boxing the int.

## API Proposal
https://github.com/dotnet/runtime/issues/28882

## Additions we'd like to make
- [x] Improved UX for nulls
- [x] Improved debugger viewing
- [x] Allow reference type values, e.g. string, to return null from `As<T>` without throwing
- [x] Don't allow setting Variant to a Variant
    The following code should print "System.String", but it prints "Azure.Value":
```csharp
    var value1 = new Value("hi");
    var value2 = new Value(value1);
    Console.WriteLine(value2.Type);
```
- [ ] Support numeric conversions, e.g.:
    If a cast from S to D works in standard C# code, it should work when S is stored in Variant.
```csharp
    var value2 = new Value(1);
    object int64 = value2.As<long>();
```

## Libraries we expect to use it
- JobRouter
- Azure.Core
  - to expand DynamicData to types other than JSON
  - possibly to [reduce the set of APIs](https://github.com/Azure/azure-sdk-for-net/pull/38266#discussion_r1302210825) on MutableJsonElement

**Comment by christothes** [comment]
Do we have any other context on the `Azure.Value` type?

**Comment by heaths** [comment]
@KrzysztofCwalina and @tg-msft asked me to consider using it for an upcoming KV release, but seems it won't make it in time. They asked me to open a tracking issue.

**Comment by annelo-msft** [comment]
Per @KrzysztofCwalina:
> The following package GA is blocked on this. We want to not ship this type and use Value instead: https://apiview.dev/Assemblies/Review/e1389ac1af984a86a079127e8093937a#Azure.Communication.JobRouter.LabelValue

**Comment by annelo-msft** [comment]
Filling in some context:
- KeyVault has shipped a [KeyVaultSettingValue](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultSettingValue.cs) type that would have used Azure.Value if it had been available.  Usage samples for that are [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Administration/samples/Sample4_UpdateSettings.md).
- Azure.Communication.JobRouter.LabelValue has some samples [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/communication/Azure.Communication.JobRouter/README.md#job-1)
- The original PR for Azure.Value is here: https://github.com/Azure/azure-sdk-for-net/pull/27487

**Comment by heaths** [comment]
`KeyVaultSettingValue` does do one thing a little differently than `Value`, and it may or may not be worth considering. To avoid polymorphic types for a single value, we recommended a `type: string` property for the value. Any primitive can encode as a string, so this avoided more complexity in the generated models we use as an HLC. Thus, `ToString` always returns the string-formatted value, and customers can set a string. This allows a customer on an older SDK version to read/write values that are added to the service, which are not part of the schema (the key/value pairs themselves).

Maybe `Value` is meant more for polymorphic types, but this might be something to consider. At least when I opened this tracking issue, `Value` did not support an everything-as-a-string fallback.
